pkgs/applications/networking/instant-messengers/discord/default.nix

```diff
@@ -43,13 +44,15 @@ stdenv.mkDerivation rec {
         ln -s $out/opt/discord/Discord $out/bin/
         ln -s $out/opt/discord/discord.png $out/share/pixmaps
 
-        ln -s "${desktopItem}/share/applications" $out/share/
+        mkdir -p $out/share/applications
+        substitute $desktopItem/share/applications/discord.desktop $out/share/applications/discord.desktop \
+          --subst-var out
         '';
 
     desktopItem = makeDesktopItem {
       name = pname;
-      exec = "Discord";
-      icon = pname;
+      exec = "@out@/bin/Discord";
+      icon = "@out@/share/pixmaps/discord.png";
```

Comment:
Setting the icon with an absolute path is going to break custom icons in themes.

Reply:
Ah damn, it currently only works like this in ChromeOs.

I believe the Slack package we use the default pixmap.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/slack/default.nix#L91

I'm not sure if this is just something that was missed though.
If this is the case, then I may start a downstream library with the desktopfiles patched to work in Crostini.

Reply:
Yeah, icon is not supposed to be a full path but many places do that anyway. You should be able to add the /share directory to XDG_DATA_DIRS. Here's someone who looks like they got it to work:

https://www.reddit.com/r/Crostini/comments/9lazhn/environment_variables_for_launcher_applications/e769u4y/

As an aside, it would be great to have someone document using Nixpkgs with Crostini on the wiki: https://nixos.wiki

Reply:
@matthewbauer I agree, I think it's a good idea to put some docs up regarding running Nix in Crostini. I have put up a very quick page (something that can be worked on and refined as things progress).

I think I need to spend some more time figuring out where Crostini is going with image recognition as there are a few limitations I have come across in this process.

- only supports .png
- can only pick up images in the pixmaps directory (ie. vlc/xterm not supported)
- links to executables should be absolute (although this now works if using the above reddit advice).

I'm not sure what the correct course of action here is, I understand the want for custom themes, so don't want to rule that out, I will play around more with it, and see if I can create a better user-story for using Nix within Crostini (it's already pretty decent tbh).


Link: https://nixos.wiki/wiki/Installing_on_Crostini